distsql/request_builder.go

```diff
@@ -255,6 +260,42 @@ func (builder *RequestBuilder) SetTiDBServerID(serverID uint64) *RequestBuilder
 	return builder
 }
 
+func (builder *RequestBuilder) verifyTxnScope(sv *variable.SessionVars) {
+	if builder.Request.TxnScope == oracle.GlobalTxnScope {
+		return
+	}
+	is := infoschema.GetInfoSchemaBySessionVars(sv)
+	for _, bundle := range is.RuleBundles() {
+		rule, ok := placement.GetLeaderRuleByBundle(bundle, placement.DCLabelKey)
+		if !ok {
+			continue
+		}
+		dc, ok := placement.GetLeaderDCByBundle(bundle, placement.DCLabelKey)
+		if !ok {
+			continue
+		}
+		ptID, err := placement.ObjectIDFromGroupID(rule.GroupID)
+		if err != nil {
+			builder.err = err
+			continue
+		}
+		intersectTableID := int64(0)
+		for _, keyRange := range builder.Request.KeyRanges {
+			tableID := tablecodec.DecodeTableID(keyRange.StartKey)
+			if ptID == tableID {
+				intersectTableID = tableID
+				break
+			}
+		}
```

Comment:
`tableID` may not exist in placement rules. In this case, an error should be reported.

Reply:
I think you should build a map to store all the `tableID` that are included in the request, then you look for the table ID from the rules. Iterating all the rules is not a good idea because there may be many rules.

Reply:
updated.